README

1.  Export the EI you want to build into /usr/local/mgi/ei/ei_build:

    Note that exporting the EI is NOT an installation of the EI!!!
    
    cd /usr/local/mgi/ei/ei_build
    cvs export -r ei-tag -d ei-tag ei

2.  Link /usr/local/mgi/ei/ei_build/ei-tag to /usr/local/mgi/ei/ei_build/ei

    rm -rf ei	(remove existing link)
    ln -s ei-tag ei

3.  Edit Configuration file:

    cd ei-tag
    cp Configuration.default Configuration
    edit Configuration file carefully and appropriately

    on rohan (development):
	setenv EIINSTALLDIR /usr/local/mgi/ei/dev
	setenv USERLOG  $HOME/.eidev-log
	setenv DSQUERY DEV_MGI
	setenv MGD mgd_release
	setenv NOMEN nomen_release

    on shire (production):
	use defaults

4.  Building

    note: install also does a build.  so, if you're ultimate
    goal is to install the application, you can skip this step.

    one application:

    	buildapp mgd compiled

    all applications:

    	buildall compiled

    For production and integration/beta testing, install compiled.
    For debugging, leave off "compiled" option.

5.  Installing

    one application:

    	installapp mgd compiled

    all applications:

    	installall compiled

    For production and integration/beta testing, install compiled.
    For debugging, leave off "compiled" option.

6.  Running:

    rohan:
    	/usr/local/mgi/ei/dev/runapp mgi

    shire:
    	/usr/local/mgi/ei/prod/runapp mgi



HISTORY
## Remember to update Configuration:CVS_TAG with most recent tag number
## Remember to update Configuration:CVS_TAG with most recent tag number

TAG: ei-6-0-17-1
DATE: 05/11/2021
STAFF: lec
TR13349/Genome Build 39

TAG: ei-6-0-16-9
DATE: 09/28/2020
STAFF: lec
Probe/ProbeSource/autosequence

TAG: ei-6-0-16-7
TAG: ei-6-0-16-6
TAG: ei-6-0-16-5
TAG: ei-6-0-16-8
DATE: 09/24/2020
STAFF: lec
PythonLib.d/change call to Python 3

TAG: ei-6-0-16-7
TAG: ei-6-0-16-6
TAG: ei-6-0-16-5
TAG: ei-6-0-16-4
TAG: ei-6-0-16-3
DATE: 09/22/2020
STAFF: lec
GXD/fix GXD_AssayNote, GXD_GelLaneStructure, GXD_ISReseultStructure, GXD_InSituResultImage
due to schema changes

TAG: ei-6-0-16-2
TAG: ei-6-0-16-1
DATE: 08/21/2020
STAFF: lec
TR13204/Infrastructure Release

TAG: ei-6-0-15-6
DATE: 03/09/2020
STAFF: lec
CHANGES: TR13278/Teluse EI strain module ADD bug
dsrc/Strains.d/Add()/ProcessAcc()

TAG: ei-6-0-15-5
TAG: ei-6-0-15-4
DATE: 01/29/2020
STAFF: lec
CHANGES: tr13007-genotye merge

TAG: ei-6-0-15-3
DATE: 01/21/2020
STAFF: lec 
Marker/Non-Mouse/dsrc/SQL.d/
removed: select * from markerKeyMax;/obsolete

TAG: ei-6-0-15-2
TAG: ei-6-0-15-1
DATE: 12/12/2019
STAFF: lec 
TR13007/RNA Sequence project

TAG: ei-6-0-14-10
DATE: 08/09/2019
STAFF: lec 
TR13169/add KMPC to Strain module

TAG: ei-6-0-14-9
DATE: 05/09/2019
STAFF: lec 
Configuration/Install changes for new bheidev01 server

TAG: ei-6-0-14-8
DATE: 05/07/2019
STAFF: lec 
TR13109/specimen note tag

TAG: ei-6-0-14-7
TAG: ei-6-0-14-6
TAG: ei-6-0-14-4
TAG: ei-6-0-14-3
TAG: ei-6-0-14-2
TAG: ei-6-0-14-1
DATE: 04/03/2019, 04/15/2019
STAFF: lec 
TR12963/teleuse-to-mgd_java_api

TAG: ei-6-0-13-6
DATE: 03/06/2019
STAFF: lec
CHANGES: TR13053/add GPT logicaldb/Strain module

TAG: ei-6-0-13-5
TAG: ei-6-0-13-4
DATE: 12/21/2018
STAFF: lec
CHANGES: 
TR12979/fix Strain/marker/allele display
TR12975/fix csrc/mgilib.c/mgiDBincKey/refStrainKey

TAG: ei-6-0-13-3
DATE: 12/20/2018
STAFF: lec
CHANGES: 
TR12978/reference/strain default to "Additional"
TR12980/readd ENVIGO to Strain

TAG: ei-6-0-13-2
TAG: ei-6-0-13-1
DATE: 09/04/2018, 09/12/2018
STAFF: lec
CHANGES: TR12734/GenFeVah merge

TAG: ei-6-0-12-9
DATE: 05/18/2018
TAG: ei-6-0-12-11
DATE: 07/31/2018
STAFF: lec
CHANGES: TR12911/add CWR logicaldb/Strain module

TAG: ei-6-0-12-9
DATE: 05/18/2018
TAG: ei-6-0-12-10
DATE: 06/08/2018
STAFF: lec
CHANGES: TR12876/ENVIGO added to ACC_LogicalDB/Strain

TAG: ei-6-0-12-9
DATE: 05/18/2018
STAFF: lec
CHANGES: TR12827/age range column

TAG: ei-6-0-12-8
DATE: 05/11/2018
STAFF: lec
CHANGES: TR12856/increase marker symbol
pcds/MGD.pcd

TAG: ei-6-0-12-7
TAG: ei-6-0-12-6
TAG: ei-6-0-12-5
TAG: ei-6-0-12-4
DATE: 05/02/2018
STAFF: lec
CHANGES:
csrc/mgilib.c : drop nextval table if it exists

TAG: ei-6-0-12-3
TAG: ei-6-0-12-2
TAG: ei-6-0-12-1
DATE: 03/07/2018, 03/12/2018
STAFF: lec
CHANGES:
TR12662/GMC project
TR12813/GXD Assay/Age Range

TAG: ei-6-0-11-10
DATE: 03/01/2018
STAFF: lec
CHANGES:
fix MGI_Organism/sequencenum bug

TAG: ei-6-0-11-9
TAG: ei-6-0-11-8
DATE: 01/23/2018
STAFF: lec
CHANGES:
TR12770/added UNC/208 to Strains/pick list

TAG: ei-6-0-11-5
DATE: 01/04/2018
TAG: ei-6-0-11-7
TAG: ei-6-0-11-6
DATE: 01/10/2018
STAFF: lec
CHANGES:
increate Allele/Symbol column to 75

TAG: ei-6-0-11-5
DATE: 01/04/2018
STAFF: lec
CHANGES:
TR12744/removing carriage returns from standardized notes

TAG: ei-6-0-11-3
DATE: 11/15/2017
TAG: ei-6-0-11-4
DATE: 11/20/2017
STAFF: lec
CHANGES:
added 207/CHROMus to Strain module

TAG: ei-6-0-11-3
DATE: 11/15/2017
STAFF: lec
CHANGES:
1)TR12662/GMC/Organism/changed query to remove hardcoded mgitypes
2)removed GXD Index module/replaced by PWI

TAG: ei-6-0-11-2
TAG: ei-6-0-11-1
DATE: 10/04/2017, 10/05/2017
STAFF: lec
CHANGES:
TR12250/Lit Triage

TAG: ei-6-0-10-9
TAG: ei-6-0-10-8
DATE: 06/08/2017
STAFF: lec
CHANGES:
TR12605/Add MirKO-project to allele module
TR12597/Reference module/allow chagne of J: for Reserved
TR12604/dsrc/MutantCellLine.d/vector type fix

TAG: ei-6-0-10-7
DATE: 05/25/2017
STAFF: lec
CHANGES:
re-installed the dsrc/Reference/ref_nom_exists

TAG: ei-6-0-10-5
TAG: ei-6-0-10-4
TAG: ei-6-0-10-6
DATE: 05/23/2017
STAFF: lec
CHANGES:
added 'genOway' per Howie

TAG: ei-6-0-10-5
TAG: ei-6-0-10-4
TAG: ei-6-0-10-3
TAG: ei-6-0-10-2
TAG: ei-6-0-10-1
DATE: 04/27/2017
STAFF: lec
CHANGES:
TR12540/Disease Ontology (DO) All The Way
1) OMIMVocAnnot.d : obsolete
2) DOAlleleDiseaseVocAnnot renamed AlleleDiseaseVocAnnot
3) TR12514/GXD Images/add Creative Commons red label

TAG: ei-6-0-8-2
TAG: ei-6-0-8-1
DATE: 10/12/2016, 02/14/2017
STAFF: lec
CHANGES:
TR12427/Disease Ontology (DO)
TR11083/Nomenclature
1) Verify.d/VerifyVocabTermAccID
2) remove DO Annotations from Genotype/MGD menus

TAG: ei-6-0-6-6
DATE: 12/14/2016
STAFF: lec
CHANGES:
dsrc/AccLib.d : changed ilike to like

TAG: ei-6-0-6-5
DATE: 12/01/2016
STAFF: lec
CHANGES:
1) added CRL - Charles River Laboratory/Strain module

TAG: ei-6-0-6-4
DATE: 11/01/2016
STAFF: lec
CHANGES:
1) TR12370/scrum-bob/MGI_Organism_MGIType schema change
2) TR12448/remove dsrc//Verify.de/Verify.d/VerifyStrainNomenclature

TAG: ei-6-0-6-3
TAG: ei-6-0-6-2
TAG: ei-6-0-6-1
DATE: 08/05/2016
STAFF: lec
CHANGES:
TR12349/gone-dor & more 
TR12345/GPAD/GPI
1) add Verify/VerifyGOIsoform
2) mgdsql.c/_marker_status_key = 3 (interim) is obsolete
3) TR12331/add MDP to Strain module
4) TR12229/add PRO to Reference module/data sets
5) TR12363/GXD Index/coded? query

TAG: ei-6-0-4-8
TAG: ei-6-0-4-7
TAG: ei-6-0-5-1
DATE: 08/03/2016
STAFF: lec
CHANGES:
1) Add SMOC/Shanghai Model Organisms Center (184) to Strain module

TAG: ei-6-0-4-8
TAG: ei-6-0-4-7
DATE: 07/22/2016
STAFF: lec
CHANGES:
1) dsrc/Reference/isIncompelte/check if variable exists

TAG: ei-6-0-4-6
DATE: 06/06/2016
STAFF: lec
CHANGES:
1) replace CVS_TAG with GITHUB_TAG

TAG: ei-6-0-4-5
DATE: ?
STAFF: lec
CHANGES:
1) TR6939/assay/genotype/replace dialog

TAG: ei-6-0-4-4
TAG: ei-6-0-4-3
TAG: ei-6-0-4-2
TAG: ei-6-0-4-1
DATE: 05/04/2016
STAFF: lec
CHANGES:
1) TR12223/gxd anatomy II : merge to trunk

TAG: ei-6-0-3-3

dsrc/SQL.d
--
-- Name    : SQL.d
-- Creator : lec
-- SQL.d 02/12/99
--
-- This modules contains events which actually execute 99% of the SQL
-- commands in the EI.
--
-- AddSQL, ModifySQL, DeleteSQL all call ExecSQL which actually executes
-- the SQL commands.  ExecSQL then processes any error messages returned
-- from the error handler defined in pglib.c.
--
-- The Query events handle all dynamic searching w/in the EI.
--
-- History
--
-- lec 03/30/2006
--	- APP_JobStream: dbsnpload is okay to run 
--
-- lec 01/10/2002
--	- QueryNoInterrupt; use LoadList
--
-- lec 09/18/2001
--	- QueryNoInterrupt; added "selectItem"
--
-- lec 09/12/2001
--	- AddSQL; add appendKeyToItem
--
-- lec 08/28/2001
--	- QueryNoInterrupt; write SQL to log
--
-- lec 03/08/2001-3/19/2001
--	- TR 2217/1939; ModifyNotes
--	- moved ModifyNotes to NoteLib.d
--	- moved QueryDate to DateLib.d
--
-- lec 08/11/99
--	- TR 812; added printSelect to Query event
--
-- lec 12/23/98
--	- Query; introduce rowcount
--
-- lec 11/16/98
--	- ModifyNotes; initialize noteWidget.sql
--
-- lec 10/16/98
--	- AddSQL.key could be a table
--
-- lec 08/25/98
--	- added QueryDate
--
-- lec	08/17/98
--	- moved SetServer and SetTitle to MGI.d/MGI.de
--
-- lec	07/23/98
--	- ModifyNotes; if notes not modified, then return
--
-- lec	07/08/98
--	- delete notes (ModifyNotes) if notes have been modified
--
-- lec	07/02/98
--	- added ModifyNotes
--
-- lec	06/29/98
--	- added global_application/global_version to SetTitle
--
-- lec	06/11/98
--	- replaced tu_printf w/ mgi_writeLog
--	- added prefix CMD: to ExecSQL log
--	- added prefix QUERY: to Query log
--
-- lec	05/22/98
--	- added "begin/commit transaction" ability to ModifySQL
--

dmodule SQL is

#include <mgilib.h>
#include <dblib.h>
#include <mgisql.h>

locals:
	top : widget;
	queryList : widget;
	newID : string;

rules:

--
-- AddSQL
--
--      Execute SQL insert command
--      Set List.sqlSuccessful attribute
--      Insert record and key into selection list
--      Redisplay appropriate record count
--
 
        AddSQL does
	  cmd : string;
	  item : string := AddSQL.item;

	  if (AddSQL.list != nil) then
	    top := AddSQL.list.top;
            AddSQL.list->List.sqlSuccessful := true;
	  end if;

	  -- Enclose insert statments within a transaction
	  -- so that upon any errors the entire transaction is aborted.
	  -- There may be some cases where enclosing statements within
	  -- a transaction is not desired.  If this is the case, the
	  -- calling event can set AddSQL.transaction = false

	  if (AddSQL.transaction) then
	    cmd := AddSQL.cmd + "select * from keyMax;\n";
	  else
	    cmd := AddSQL.cmd;
	  end if;

	  ExecSQL.cmd := cmd;
	  ExecSQL.list := AddSQL.list;
	  send(ExecSQL, 0);

	  -- If no key value exists, assign it

	  if (not mgi_tblIsTable(AddSQL.key)) then
	    if (AddSQL.key.value.length = 0 and newID.length > 0) then
	      AddSQL.key.value := newID;
	    end if;
	  else
	    if (mgi_tblGetCell(AddSQL.key, AddSQL.row, AddSQL.column) = "" and newID.length > 0) then
	      (void) mgi_tblSetCell(AddSQL.key, AddSQL.row, AddSQL.column, newID);
	    end if;
	  end if;

	  -- If no list given, return

	  if (AddSQL.list = nil) then
	    return;
	  end if;

	  -- If Add was successful, 
	  --   Set key value to newID
	  --   Add entry to List
	  --   Re-count records

          if (AddSQL.list->List.sqlSuccessful) then
            if (AddSQL.appendKeyToItem) then
              item := "*[" + AddSQL.key.value + "]:  " + item;
            end if; 

            if (AddSQL.useItemAsKey) then
              InsertList.key := item;
            else
              InsertList.key := AddSQL.key.value;
            end if; 

            InsertList.list := AddSQL.list;
            InsertList.item := item;
            send(InsertList, 0); 
            top->RecordCount->text.value := mgi_DBrecordCount(AddSQL.tableID);
            (void) XmListSelectPos(AddSQL.list->List, AddSQL.list->List.row, AddSQL.selectNewListItem);
          end if; 
        end does;
 
--
-- DeleteSQL
--
--      Delete row from database
--      Delete appropriate row from selection list
--      Redisplay appropriate record count
--
 
        DeleteSQL does
          top := DeleteSQL.list.top;

          if (DeleteSQL.key2 != nil) then
             ExecSQL.cmd := mgi_DBdelete2(DeleteSQL.tableID, DeleteSQL.key, DeleteSQL.key2);
             ExecSQL.list := DeleteSQL.list;
             send(ExecSQL, 0); 
          else
             ExecSQL.cmd := mgi_DBdelete(DeleteSQL.tableID, DeleteSQL.key);
             ExecSQL.list := DeleteSQL.list;
             send(ExecSQL, 0); 
          end if; 

          -- If delete was successful, delete row from list and re-count records

          if (DeleteSQL.list->List.sqlSuccessful) then
            DeleteList.list := DeleteSQL.list;
            send(DeleteList, 0); 
            top->RecordCount->text.value := mgi_DBrecordCount(DeleteSQL.tableID);
            if (top->Control->Delete != nil) then
              top->Control->Delete.deleteReturn := true;
            end if; 
          end if; 
        end does;

--
-- ExecSQL
--
--	Execute non-null SQL command
--	Set List.sqlSuccessful attribute
--

	ExecSQL does
	  error : integer;
	  transtate : integer;

	  if (ExecSQL.list != nil) then
	    ExecSQL.list->List.sqlSuccessful := true;
	  end if;

          if (ExecSQL.cmd.length = 0) then
            return;
          end if;
 
	  if (ExecSQL.logOnly) then
	    (void) mgi_writeLog("\nturn logOnly = true\n\n");
	    return;
 	  end if;

	  -- Execute cmd

	  newID := "";
	  dbproc : opaque;
	  
	  dbproc := mgi_dbexec(ExecSQL.cmd);
          while (mgi_dbresults(dbproc) != NO_MORE_RESULTS) do
            while (mgi_dbnextrow(dbproc) != NO_MORE_ROWS) do
	      newID := mgi_getstr(dbproc, 1);
	    end while;
	  end while;
	  mgi_dbclose(dbproc);

	  -- Fatal Errors

	  (void) mgi_writeLog("fatal error : " + (string) global_error + "\n");

	  if (global_error = 1) then
	    newID := "";
	    if (ExecSQL.list != nil) then
	      ExecSQL.list->List.sqlSuccessful := false;
	    end if;
	  end if;

	end does;

--
-- ModifySQL
--
--	Execute non-null SQL command
--	Set List.sqlSuccessful attribute
--	Re-select record
--

	ModifySQL does
	  cmd : string;

	  if (ModifySQL.list != nil) then
	    ModifySQL.list->List.sqlSuccessful := true;
	    top := ModifySQL.list.root;
	  else
	    top := ModifySQL.source_widget.root;
	  end if;

	  if (ModifySQL.cmd.length = 0) then
            StatusReport.source_widget := top;
            StatusReport.message := "No Values Were Modified";
            send(StatusReport);

	    if (ModifySQL.list != nil) then
              (void) XmListSelectPos(ModifySQL.list->List, ModifySQL.list->List.row, true);
	    end if;
	    return;
	  end if;

	  -- Enclose insert statments within a transaction
	  -- so that upon any errors the entire transaction is aborted.
	  -- There may be some cases where enclosing statements within
	  -- a transaction is not desired.  If this is the case, the
	  -- calling event can set ModifySQL.transaction = false

	  cmd := ModifySQL.cmd;

	  --if (ModifySQL.transaction) then
	    --cmd := "begin transaction;\n" + ModifySQL.cmd + "\ncommit transaction;\n";
	  --else
	    --cmd := ModifySQL.cmd;
	  --end if;

	  ExecSQL.cmd := cmd;
	  ExecSQL.list := ModifySQL.list;
	  ExecSQL.logOnly := ModifySQL.logOnly;
	  send(ExecSQL, 0);
 
	  if (top.is_defined("allowSelect") != nil) then
	    top.allowSelect := true;
	  end if;

	  -- Re-select record

	  if (ModifySQL.list != nil and ModifySQL.reselect) then
            (void) XmListSelectPos(ModifySQL.list->List, ModifySQL.list->List.row, true);
	  end if;

	end does;

--
-- Query - copy of QueryNoInterrupt
--
--	Perform database query w/out allowing interuption
--	Store results in QueryList
--

        Query does
          list_w : widget;

          if (Query.list_w = nil) then
            list_w := Query.source_widget->QueryList;
          else  
            list_w := Query.list_w;
          end if;

	  -- Clear List before performing new search

          ClearList.source_widget := list_w;
          send(ClearList, 0);

	  --(void) mgi_writeLog(get_time() + "QUERY:" + Query.select + "\n");

	  list_w.cmd := Query.select;
	  LoadList.list := list_w;
	  send(LoadList, 0);

	  top := list_w.top;
	  if (top.is_defined("allowSelect") != nil) then
	    top.allowSelect := true;
	  end if;

	  if (list_w->List.itemCount > 0) then
            list_w->List.row := 1;

	    --if (Query.selectItem) then
            (void) XmListSelectPos(list_w->List, list_w->List.row, true);
	    --end if;
	  end if;

	  (void) mgi_writeLog("\n\nRESULTS COUNT: " + (string) list_w->List.itemCount + "\n\n");
	  (void) reset_cursor(Query.source_widget);

        end does;
 
--
-- QueryNoInterrupt
--
--	Perform database query w/out allowing interuption
--	Store results in QueryList
--

        QueryNoInterrupt does
          list_w : widget;

          if (QueryNoInterrupt.list_w = nil) then
            list_w := QueryNoInterrupt.source_widget->QueryList;
          else  
            list_w := QueryNoInterrupt.list_w;
          end if;

	  -- Clear List before performing new search

          ClearList.source_widget := list_w;
          send(ClearList, 0);

	  --(void) mgi_writeLog(get_time() + "QUERY:" + QueryNoInterrupt.select + "\n");

	  list_w.cmd := QueryNoInterrupt.select;
	  LoadList.list := list_w;
	  send(LoadList, 0);

	  top := list_w.top;
	  if (top.is_defined("allowSelect") != nil) then
	    top.allowSelect := true;
	  end if;

	  if (list_w->List.itemCount > 0) then
            list_w->List.row := 1;

	    if (QueryNoInterrupt.selectItem) then
              (void) XmListSelectPos(list_w->List, list_w->List.row, true);
	    end if;
	  end if;

	  (void) mgi_writeLog("\n\nRESULTS COUNT: " + (string) list_w->List.itemCount + "\n\n");
	  (void) reset_cursor(QueryNoInterrupt.source_widget);

        end does;
 
--
-- StatusReport
--
-- Display Status Report to user
--

        StatusReport does
	  status : widget;

	  if (StatusReport.source_widget != nil) then
	    status := StatusReport.source_widget;
	  else
	    status := top;
	  end if;

	  -- Do not overwrite Status Dialog if already managed

          if (status->StatusDialog = nil) then
	    --(void) mgi_writeLog(get_time() + "ERROR: Could not get StatusDialog\n");
	    return;
	  end if;

          if (not status->StatusDialog.managed) then
            status->StatusDialog.messageString := StatusReport.message;
            status->StatusDialog.managed := true;
	    XmUpdateDisplay(status->StatusDialog);
	  elsif (StatusReport.appendMessage = true) then
            status->StatusDialog.managed := false;
            status->StatusDialog.messageString := 
	    	status->StatusDialog.messageString + "\n\n" + StatusReport.message;
            status->StatusDialog.managed := true;
	  end if;

          status->StatusDialog.top.front;
        end does;

--
-- StatusReportOK
--
-- Special callback for upper level Status Report dialog
-- After unmanaging dialog, place Menu shell in back of stacking order
--

        StatusReportOK does
--	  if (StatusReportOK.source_widget = top->StatusDialog and
--	      top.name != "Login") then
--	    top.back;
--	  end if;

	  StatusReportOK.source_widget.managed := false;
          StatusReportOK.source_widget.messageString := "";
	end does;

end dmodule;
